rust/datafusion/src/physical_plan/hash_utils.rs

```diff
@@ -0,0 +1,144 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Functionality used both on logical and physical plans
+
+use crate::error::{DataFusionError, Result};
+use arrow::datatypes::{Field, Schema};
+use std::collections::HashSet;
+
+/// All valid types of joins.
+#[derive(Clone, Debug)]
+pub enum JoinType {
+    /// Inner join
+    Inner,
+}
+
+/// Checks whether the schemas "left" and "right" and columns "on" represent a valid join.
+/// They are valid whenever their columns' intersection equals the set `on`
+pub fn check_join_is_valid(
+    left: &Schema,
+    right: &Schema,
+    on: &HashSet<String>,
+) -> Result<()> {
+    let left: HashSet<String> = left.fields().iter().map(|f| f.name().clone()).collect();
+    let right: HashSet<String> =
+        right.fields().iter().map(|f| f.name().clone()).collect();
+
+    check_join_set_is_valid(&left, &right, &on)
+}
+
+/// Checks whether the sets left, right and on compose a valid join.
+/// They are valid whenever their intersection equals the set `on`
+fn check_join_set_is_valid(
+    left: &HashSet<String>,
+    right: &HashSet<String>,
+    on: &HashSet<String>,
+) -> Result<()> {
+    if on.len() == 0 {
+        return Err(DataFusionError::Plan(
+            "The 'on' clause of a join cannot be empty".to_string(),
+        ));
+    }
+
+    let on_columns = on.iter().collect::<HashSet<_>>();
+    let common_columns = left.intersection(&right).collect::<HashSet<_>>();
+    let missing = on_columns
+        .difference(&common_columns)
+        .collect::<HashSet<_>>();
+    if missing.len() > 0 {
+        return Err(DataFusionError::Plan(format!(
+                "The left or right side of the join does not have columns {:?} columns on \"on\": \nLeft: {:?}\nRight: {:?}\nOn: {:?}",
+                missing,
+                left,
+                right,
+                on,
+            ).to_string()));
+    };
+    Ok(())
+}
+
+/// Creates a schema for a join operation.
+/// The fields "on" from the left side are always first
+pub fn build_join_schema(
+    left: &Schema,
+    right: &Schema,
+    on: &HashSet<String>,
+    join_type: &JoinType,
+) -> Schema {
+    let fields: Vec<Field> = match join_type {
+        JoinType::Inner => {
+            // inner: all fields are there
+
+            let on_fields = left.fields().iter().filter(|f| on.contains(f.name()));
+
+            let left_fields = left.fields().iter().filter(|f| !on.contains(f.name()));
+
+            let right_fields = right.fields().iter().filter(|f| !on.contains(f.name()));
+
+            // "on" are first by construction, then left, then right
+            on_fields
+                .chain(left_fields)
+                .chain(right_fields)
+                .map(|f| f.clone())
+                .collect()
```

Comment:
It would be more standard to have all the left fields followed by all the right fields. Also, the new schema should have fully qualified names that include the table/alias prefix.

For example if we join table 'a' with fields `id` and name` with table 'b' that also has fields `id` and `name`, the output schema would be `a.id`, `a.name`, `b.id`, `b.name`.

It is quite possible that we are lacking support currently for being able to reference these qualified field names, so we will need to deal with that before we can fully integrate this new join operator into DataFusion.

Reply:
Well spotted.

Since expressions may now contain columns from multiple Schemas/RecordBatch, I think that we need to modify all signatures of `PhysicalExpr` to receive multiple `Schema` and `RecordBatch`, as expressions may now contain columns from more than one `RecordBatch`.

I agree that this is a blocker. I will work on that and leave this PR for now, as we can't progress any further here.

Reply:
I don't understand why we would want to do that. Each operator has a single output schema. The join operator creates a new schema derived from the input schemas.

Reply:
The output schema of the join would simply have fields named "a.id", "a.name", "b.id", "b.name" based on the previous example.

Reply:
@jorgecarleitao Just wanted to make sure you saw this :point_up:

Reply:
You are right. I am not thinking strait. 

My (incorrect) concern was that expressions of the form `UPPER(a.name) = UPPER(CONCAT(b.first, ' ', b.last))` would need to be represented as `PhysicalExpr`, but I totally forgot that the `=` there is _not_ the equal operator, but the "when equal hash", and therefore that expression does not mean "left equals right", that a physical expression would encapsulate.

Reply:
> For example if we join table 'a' with fields id and name with table 'b' that also has fields id and name, the output schema would be a.id, a.name, b.id, b.name.

In general the nodes are not named, which means that `'a'` is arbitrary. Spark DataFrames do not append the qualifiers, as nodes are usually not named (`df.select('a', 'b').join(df1.select('a', 'c'), on='a')` returns columns `'a','b','c'`). 

What I dislike in Spark is that it allows two columns to have the same name, i.e. `df.select('a', 'b').join(df1.select('a', 'b'), on='a')` returns columns `'a','b','b'`, but if we then write `.select('b')`, it raises.

The current implementation uses what I consider an ideal implementation: no qualifiers are added, but it raises during the planning of the join if two columns will end up with the same name (as we do for other nodes). I.e. this would be possible:

```rust
df
    .select('a', 'b')
    .join(df1, on=col("a", 0) == col("a", 1))
    .select(vec![col('a')])
```

If we introduce qualifiers, we need to use something like

```rust
df
    .select('a', 'b')
    .join(df1, on=on_col("a", 0) == on_col("a", 1))
    .select(vec![col('?.b')])
```

where `?` is an arbitrary qualifier (`1?`, `t1`? `a`?). Moreover, that qualifier will depend on whether `df1` has a column named `b` or not, which IMO leads to surprising results (the resulting schema may be `?.b` or `b` depending on whether the nodes have colliding column names).

The current code implements the principle of least surprise together with the "explicit is better than implicit": columns are named according to what the user set them to be named, and it is the user's responsibility to ensure that two column names do not collide in a join (i.e. DataFusion errors with a nice message).

Reply:
Note that I am looking at this from the optic of the DataFrame API. For SQL I imagine that `a.b` is prevalent because that is how in SQL we refer to a column of a specific table `a` (as all tables are named).

Reply:
I think you're right. This is simpler and forces the user to add aliases as required to avoid conflicts.

Reply:
I do think we need to support the case where the join key columns have different names in each table though. Here is an example from one of the simplest TPC-H queries.

```sql
FROM
    orders,
    lineitem
WHERE
    o_orderkey = l_orderkey
```

Reply:
Good point. I will prototype to double check that it works well at the schema validation.

Just to check whether we agree on some cases:

```
df.columns = ['a', 'l_orderkey']

df1.columns = ['o_orderkey', 'l_orderkey']

df.join(df1, on=df['l_orderkey'] == df1['o_orderkey'])

# raises an error (there will be two columns named `l_orderkey`).
```

```
df.columns = ['l_orderkey', 'b']

df1.columns = ['o_orderkey', 'c']

df.join(df1, on=df['l_orderkey'] == df1['o_orderkey'])

# Ok: columns = ['l_orderkey', 'b', 'o_orderkey', 'c']  (order: left, right, different column names, both are returned)
```

```
df.columns = ['a', 'o_orderkey']

df1.columns = ['o_orderkey', 'c']

df.join(df1, on=df['o_orderkey'] == df1['o_orderkey'])

# Ok: columns = ['a', 'o_orderkey', 'c']  (order: left, right, same column name, only one is returned)
```

Reply:
I don't think we should error (long term) if the two inputs have the same column names. It is fine for an initial  implementation, but I think we should automatically created aliases to disambiguate such outputs long term

Reply:
@alamb , it is difficult to argue against no arguments :)

My argument is based on the principle of least surprise: every time we desambiguate, we surprise the users by mutating the finely crafted plan that they created.

Note that this does not happen in SQL, as in SQL all names are disambiguated by the table qualifier (e.g. `SELECT a.b, b.b ...`)